src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
+
+            // DFS. If we encounter any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            while let Some(decendant) = visit_next.pop_back() {
+                if visited.is_visited(&decendant) {
+                    continue;
+                }
+
+                for edge in graph.edges(decendant) {
+                    if nodes.contains(&edge.target()) {
+                        // we found a path back to `nodes`
+                        return false;
+                    }
+
+                    visit_next.push_back(edge.target());
+                }
+
+                visited.visit(decendant);
+            }
+
+            true
+        }
+
+        macro_rules! assert_node_exists {
+            ($node_index:ident) => {
+                if self.graph.node_weight($node_index).is_none() {
+                    return Err(PyIndexError::new_err(format!(
+                        "Specified node {} is not in this graph",
+                        $node_index.index()
+                    )));
+                }
+            };
+        }
+
+        if to_replace.is_empty() {
+            return Err(PyValueError::new_err("to_replace must not be empty"));
+        }
+
+        let node_index = NodeIndex::new(node);
+        assert_node_exists!(node_index);
+
+        if self.graph.neighbors_undirected(node_index).next().is_some() {
+            return Err(PyValueError::new_err(
+                "replacement node must not have connections",
+            ));
+        }
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            HashSet::with_capacity(to_replace.capacity());
+
+        for node in to_replace {
+            let index = NodeIndex::new(node);
+            assert_node_exists!(index);
+            indices_to_remove.insert(index);
+        }
+
+        if (force_check_cycle == Some(true) || self.check_cycle)
+            && !can_contract(&self.graph, &indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Substitution would create cycle(s)",
+            ));
+        }
```

Comment:
I'm not sure this is quite right actually, shouldn't this be something like:

```suggestion
        if (force_check_cycle == Some(true) || (self.check_cycle && force_check_cycle != Some(false))
            && !can_contract(&self.graph, &indices_to_remove)
        {
            return Err(DAGWouldCycle::new_err(
                "Substitution would create cycle(s)",
            ));
        }
```

because if the user explicitly sets `force_check_cycle` to `false` we want that to override the value of `self.check_cycle`.

Reply:
Originally, I was using something like `check_cycle.unwrap_or(self.check_cycle)` to get exactly that behavior, but `_add_edge` will actually still find cycles when connecting in the new node, and that is connected to `self.check_cycle`, so it'd be possible to end up in a bad state with a modified DAG. The simplest workaround I figured would be to rename the `check_cycle` param to `force_check_cycle` so the user can choose to opt-in if cycle check is disabled at the class level, but cannot opt out if it's not.

Reply:
Ah ok that makes sense, I didn't think of that. If we can't introduce a cycle if `self.cycle_check` is `true` regardless of this flag then I think we only need a `bool` here and not `Option<bool>`. Since we only are controlling whether we override this on the `self.cycle_check == false` branch. The `Option<bool>` is only needed if we wanted to override the both the true and false case.

Reply:
I think we should probably leave this as a separate pre-check so that if a cycle were to be introduced we don't potentially add edges before we encounter the cycle. Doing the cycle checking explicitly up front means we never end up with a partial modified graph.

Reply:
Yeah, I agree, cycle-checking should be done if requested as a pre-check before we modify the graph. But at the same time we should avoid doing cycle-checks again in `_add_edge` if for example `force_check_cycle == true && self.check_cycle == true`.

Reply:
I think what @georgios-ts is suggesting is that we add a `force_cycle_check` parameter to `_add_edge` as well so that we can skip the check if `substitute_nodes_with_node` is called with `force_cycle_check=false`. We would still do the check up front in `substitute_nodes_with_node` if the caller specified `force_cycle_check=true`.

In that case, I would also rename `force_cycle_check` back to `cycle_check`, since it then gives the user the power to turn off cycle checking for just this call, irrespective of the class setting.

Reply:
Done in 4d31566. I split `_add_edge` into two methods, moving the actual adding to `_add_edge_no_checks`. Then, we always call `_add_edge_no_checks` from `substitute_nodes_with_node` since checks are done up front, and we can avoid redoing them this way. I considered adding a parameter to `_add_edge` instead, but there's no function overloading or optional parameters in Rust, so I would have had to litter all ~18 callsites with `None`s.